framework.go

```diff
@@ -0,0 +1,171 @@
+package clustertest
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/giantswarm/clustertest/pkg/application"
+	"github.com/giantswarm/clustertest/pkg/client"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	applicationv1alpha1 "github.com/giantswarm/apiextensions-application/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	cr "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	KubeconfigEnvVar = "E2E_KUBECONFIG"
+)
+
+// Framework is the overall framework for testing of clusters
+type Framework struct {
+	DisableLogging bool
+
+	kubeconfigPath string
+	mcClient       *client.Client
+	wcClients      map[string]*client.Client
+}
+
+// New initializes a new Framework instance using the kubeconfig found in the env var `E2E_KUBECONFIG`
+func New() (*Framework, error) {
+	mcKubeconfig, ok := os.LookupEnv(KubeconfigEnvVar)
+	if !ok {
+		return nil, fmt.Errorf("no %s set", KubeconfigEnvVar)
+	}
+
+	framework := NewWithKubeconfig(mcKubeconfig)
+
+	// for _, envVar := range os.Environ() {
+	// 	if strings.HasPrefix(envVar, "E2E_WC_") {
+	// 		// TODO: Initialize workload cluster
+	// 	}
+	// }
+
+	return framework, nil
+}
+
+// NewWithKubeconfig generates a new framework initialised with the Management Cluster provided as a Kubeconfig
+func NewWithKubeconfig(kubeconfigPath string) *Framework {
+	mcClient, err := client.New(kubeconfigPath)
+	if err != nil {
+		panic(err)
+	}
+
+	return &Framework{
+		kubeconfigPath: kubeconfigPath,
+		mcClient:       mcClient,
+		wcClients:      map[string]*client.Client{},
+	}
+}
+
+// MC returns an initialized client for the Management Cluster
+func (f *Framework) MC() *client.Client {
+	return f.mcClient
+}
+
+// WC returns an initialized client for the Workload Cluster matching the given name.
+// If no Workload Cluster is found matching the given name an error is returned.
+func (f *Framework) WC(clusterName string) (*client.Client, error) {
+	c, ok := f.wcClients[clusterName]
+	if !ok {
+		return nil, fmt.Errorf("workload cluster not found for name %s", clusterName)
+	}
+	return c, nil
+}
+
+// ApplyCluster takes a Cluster object, applies it to the MC in the correct order and then waits for a valid Kubeconfig to be available
+//
+// A timeout can be provided via the given `ctx` value by using `context.WithTimeout()`
+func (f *Framework) ApplyCluster(ctx context.Context, cluster *application.Cluster) (*client.Client, error) {
```

Comment:
I'm using kubectl-gs (code) to generate the Apps and Configmaps. It was explicitly decided not to exec out to k-gs because that's just one of the ways our customers interact with our product. Our actual contract is the pairing of an App CR with a values ConfigMap. E.g. gitops, happa and kubectl-gs all end up producing this but in different ways.